ES-Module support

ctavan opened the pull request:
Discussion: #330

ctavan commented:
@43081j @DarylThayil since you have commented on #317: Would you mind testing out my development branch from this PR by installing the following package directly from github:

```
  "dependencies": {
    "uuid": "github:ctavan/uuid-esm#master"
  }
```

You can have a look at some examples of how this development version can be used with esmodules in https://github.com/kelektiv/node-uuid/tree/c1307d3fde47d2191bee36d5ea2f42310c63fef8/examples

Would be curious about your feedback!

broofa commented on src/lib/v35.js:
```diff
@@ -49,9 +52,9 @@ module.exports = function(name, version, hashfunc) {
-  generateUUID.URL = '6ba7b811-9dad-11d1-80b4-00c04fd430c8';
+  // For CommonJS default export support
+  generateUUID.DNS = DNS;
+  generateUUID.URL = URL;
```
Maybe have the v3/v5.js files explicitly `export DNS, URL from `./lib/v35.js` rather than bodging it in here?  (Would that work?)

ctavan replied:
I tried that but it does not seem to work.

I'm using https://www.npmjs.com/package/babel-plugin-add-module-exports to add `module.exports` to the transpiled CommonJS modules so people can continue to do

```js
const uuidv4 = require('uuid/v4');
```

instead of having to do ugly workarounds like

```js
const uuidv4 = require('uuid/v4').default;
```

or

```js
const { default: uuidv4 } = require('uuid/v4');
```

As soon as there is more than `export default …` this `module.exports` approach will no longer work.

ctavan commented on examples/browser-esmodules/example.js:
```diff
@@ -0,0 +1,34 @@
+
+console.log('uuidv1: ', uuidv1());
+
+
```
I would even consider to add `prettier` to get rid of any formatting discussion. However since this PR is already getting really big I wanted to push these kinds of tooling changes and refactorings to a different PR. Fine for you?

ctavan commented on examples/browser-esmodules/example.html:
```diff
@@ -0,0 +1,3 @@
+<!DOCTYPE html>
+<title>UUID esmodule example</title>
+<script type="module" src="example.js"></script>
```
Will add explanation to the files.

How do you feel about adding these examples at all? Helpful or not?

ctavan commented:
@broofa since I pushed so many fixups I decided to squash all my fixups together again.

Quick summary of my changes:

* moved the `./lib` code up one level
* as per #330 went for CommonJS-build for node and esm build for browsers (webpack/rollup/etc. through `module:` and `browser:` properties in `package.json`
* as discussed in #330 decided not to change the module API at this point
* instead, removed the deprecation for the default import and instead adjusted the examples to use `const uuid = require('uuid');` again -> browser builds where module size matters should now make use of esm!
* made `v4` the first quickstart example to work against the above-the-fold-v1-bias

Here's what the contents of the npm tarball would look like: https://github.com/ctavan/uuid-esm (you can also check out the updated README there)

Feature-wise I would love to merge this state into a `next` branch and make a `uuid@4.0.0-alpha.1` prerelease.

I would then like to make some more tooling improvements (like publishing changelog to the Github releases page), maybe add browser tests (now that we have code specifically targeting the browser) and remove some outdated code [like pre-node-4 Buffer code](https://github.com/kelektiv/node-uuid/blob/eea07e8484f72e52ba935c1b9cb491a2365ad858/lib/sha1.js#L14), but all of that in smaller, separate PRs.

broofa commented:
This all sounds good.  I like the idea of a `next` branch.  Given that this is a top-20 NPM module, letting this simmer for a bit so we get more eyeballs on it is probably wise.  It might be worth adding a header to the README on `master` informing people about this branch so they can weigh in.

I'd like to keep whatever we're doing here in sync with https://github.com/tc39/proposal-uuid, whatever that means.  At a minimum, try to have some agreement on how the two projects interrelate would be prudent.  E.g. Do we treat this module as the "experimental" sandbox for whatever API(s) might come later there?  Dunno.

Note that your work here supersedes much of my `es6-modules` branch.  We should probably just delete that branch to avoid confusion.  My only hesitatio there is that the [UUID class](https://github.com/kelektiv/node-uuid/blob/es6_modules/UUID.js) has some useful code in it for parsing, validation, and working with byte arrays.  Part of me wants to say we should incorporate that, but I'm not sure I can make a good enough argument for that.

I'm traveling this week and next, but I'll try to review again ASAP.

ctavan commented:
> This all sounds good. I like the idea of a `next` branch. Given that this is a top-20 NPM module, letting this simmer for a bit so we get more eyeballs on it is probably wise. It might be worth adding a header to the README on `master` informing people about this branch so they can weigh in.

👍 

I would need publish access for the npm module at some point though to publish `4.0.0-alpha` prereleases.

Will make sure to update the master README.

> I'd like to keep whatever we're doing here in sync with https://github.com/tc39/proposal-uuid, whatever that means. At a minimum, try to have some agreement on how the two projects interrelate would be prudent. E.g. Do we treat this module as the "experimental" sandbox for whatever API(s) might come later there? Dunno.

👍 

However as discussed in #330 I will focus on ESModule interoperability and tooling changes for now and I will not introduce any API changes for the time being.

Should we decide to change the API of this module though, I totally agree that we should keep it in sync with https://github.com/tc39/proposal-uuid.

> Note that your work here supersedes much of my `es6-modules` branch. We should probably just delete that branch to avoid confusion. My only hesitatio there is that the [UUID class](https://github.com/kelektiv/node-uuid/blob/es6_modules/UUID.js) has some useful code in it for parsing, validation, and working with byte arrays. Part of me wants to say we should incorporate that, but I'm not sure I can make a good enough argument for that.

In fact [I cherry-picked your commit from the other branch](https://github.com/kelektiv/node-uuid/pull/331/commits/9c249a268fc5e1c84fe84c0b705caa83c1e8c14c), I see value in leaving the other branch open for a while until we have made decisions around the API changes (which I intentionally left out here).

ctavan commented:
Closing in favor of #337 337
